pkg/metrics/collector.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2018 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace       = "ndm"
+	defaultEnabled  = true
+	defaultDisabled = false
+)
+
+var (
+	factories      = make(map[string]func() (Collector, error))
+	collectorState = make(map[string]*bool)
+)
+
+// Collector is the interface a collector has to implement.
+type Collector interface {
+	// Get new metrics and expose them via prometheus registry.
+	Update(ch chan<- prometheus.Metric) error
+}
+
+type typedDesc struct {
+	desc      *prometheus.Desc
+	valueType prometheus.ValueType
+}
+
+// NodeCollector implements the prometheus.Collector interface.
+type nodeCollector struct {
+	Collectors map[string]Collector
+}
+
+func registerCollector(collector string, isDefaultEnabled bool, factory func() (Collector, error)) {
+	collectorState[collector] = &isDefaultEnabled
+	factories[collector] = factory
+}
+
+// NewNodeCollector creates a new NodeCollector
+func NewNodeCollector(filters ...string) (*nodeCollector, error) {
+	f := make(map[string]bool)
+	for _, filter := range filters {
```

Comment:
please add comments about, what is filters

Reply:
Currently there is one collector only so is use of filters. I will remove that.